### PR TITLE
bazel: docs for adding/maintaining envoy build rules (#415).

### PR DESCRIPTION
**Pull request by htuch**

### CODE REVIEW

File: source/server/config/BUILD [hunk 1]
```diff
@@ -1,3 +0,0 @@
-package(default_visibility = ["//visibility:public"])
```
Comment (mattklein123) [inline source/server/config/BUILD]: Did you mean for this to be in this change?

**Comment by htuch** [comment]
Yeah, just a cleanup I snuck in that I noticed while writing the docs.

File: bazel/DEVELOPER.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+rules](../bazel/envoy_build_system.bzl) are provided below. The [Bazel BUILD
+Encyclopedia](https://bazel.build/versions/master/docs/be/overview.html)
+provides further details regarding the underlying rules.
+
+## Style guide
+
+The [BUILD file style
+guide](https://bazel.build/versions/master/docs/skylark/build-style.html) is the
+canonical style reference. The
+[buildifier](https://github.com/bazelbuild/buildifier) tool automatically
```
Comment (mattklein123) [inline bazel/DEVELOPER.md]: Can we run this in CI like we do for check_format?

Comment (htuch) [inline bazel/DEVELOPER.md]: Good idea, I have PRs out for this in #677 and #678.

File: bazel/README.md [hunk 3]
```diff
@@ -62,3 +62,7 @@ the units tests in
 ```
 bazel test //test/commont/http:async_client_impl_test
```
Comment (hennna) [inline bazel/README.md]: Small typo: commont -> common

File: bazel/DEVELOPER.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    hdrs = ["mocks.h"],
+    deps = [
+        "//include/envoy/foo:bar_interface",
+        ...
+    ],
+)
+```
+
+Typically, mocks are provided for all interfaces in a directory in a single
+`mocks.{cc,h}` and corresponding `_mocks` Bazel target.
```
Comment (junr03) [inline bazel/DEVELOPER.md]: Not sure if it is worth it to mention different cases like https://github.com/lyft/envoy/blob/master/test/mocks/upstream/BUILD, where there are mocks in files other than mocks.{cc,h}

**Pull request merged**